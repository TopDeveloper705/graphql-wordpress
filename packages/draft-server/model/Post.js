import DataLoader from 'dataloader';
import findByIds from 'mongo-find-by-ids';
import { getUniqueSlug } from './utils';

export default class Post {
  constructor(context) {
    this.context = context;
    this.collection = context.db.collection('post');
    this.tags = context.db.collection('tag');
    this.loader = new DataLoader(ids => findByIds(this.collection, ids));
    this.tagLoader = new DataLoader(ids => findByIds(this.tags, ids));
  }

  findTags(tags) {
    return this.tagLoader.loadMany(tags);
  }

  findOneById(id) {
    return this.loader.load(id);
  }

  findOneBySlug(slug) {
    return this.collection.findOne({
      slug,
    });
  }

  count(args = {}) {
    const criteria = Object.assign({}, args);
    delete criteria.search;
    if (args.search) {
      criteria.$text = { $search: args.search };
    }
    return this.collection.find(criteria).count();
  }

  all({ limit = 10, offset = 0, search = null }) {
    const criteria = {};
    if (search) {
      criteria.$text = { $search: search };
    }

    return this.collection
      .find(criteria)
      .sort({ createdAt: -1 })
      .skip(offset)
      .limit(limit)
      .toArray();
  }

  async insert(doc) {
    const slug = await getUniqueSlug(this.collection, doc.title);
    const docToInsert = Object.assign({}, doc, {
      slug,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });
    if (!docToInsert.tags) {
      docToInsert.tags = [];
    }
    const id = (await this.collection.insertOne(docToInsert)).insertedId;
    return id;
  }

  async updateById(id, doc) {
    const docToUpdate = Object.assign({}, doc);
    const ret = await this.collection.update(
      { _id: id },
      {
        $set: Object.assign({}, docToUpdate, {
          updatedAt: Date.now(),
        }),
      }
    );
    this.loader.clear(id);
    return ret;
  }

  async removeById(id) {
    const ret = this.collection.remove({ _id: id });
    this.loader.clear(id);
    return ret;
  }
}
