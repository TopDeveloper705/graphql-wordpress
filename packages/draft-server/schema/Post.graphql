type Post {
  id: ObjID!
  title: String
  slug: String
  content: String
  contentState: ContentState
  tags: [Tag]
  createdAt: Float!
  updatedAt: Float!
}

type PostEdge {
  node: Post
  cursor: String
}

type PostConnection {
  tags: [Tag]
  count: Int
  edges: [PostEdge]
  pageInfo: PageInfo
}

extend type Query {
  posts(
    first: Int
    after: String
    last: Int
    before: String
    year: Int
    search: String
    tags: String
  ): PostConnection
  post(id: ObjID, slug: String): Post
}

input CreatePostInput {
  title: String
  contentState: ContentStateInput
  tags: [String]
}

input UpdatePostInput {
  title: String
  contentState: ContentStateInput
  tags: [String]
}

extend type Mutation {
  createPost(input: CreatePostInput!): Post
  updatePost(id: ObjID!, input: UpdatePostInput!): Post
  removePost(id: ObjID!): Boolean
}
