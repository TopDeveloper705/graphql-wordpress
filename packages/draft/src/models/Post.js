import { ObjectId } from 'mongodb';
import Model from './Model';
import { getUniqueSlug } from './utils';

function convertEntityData(entityMap) {
  return entityMap.map(entity => {
    const normalized = {};
    if (entity.data.type === 'EMBED') {
      const { type, url, html } = entity.data;
      normalized.data = { type, url, html };
    } else if (entity.data.type === 'LINK') {
      const { type, href, target } = entity.data;
      normalized.data = { type, href, target };
    } else if (entity.data.type === 'IMAGE') {
      const { type, imageId, size } = entity.data;
      normalized.data = { type, imageId, size };
      normalized.data.imageId = ObjectId(imageId);
    } else if (entity.data.type === 'VIDEO') {
      const { type, videoId } = entity.data;
      normalized.data = { type, videoId };
      normalized.data.videoId = ObjectId(videoId);
    } else {
      normalized.data = entity.data;
    }
    const { type, mutability } = entity;
    normalized.type = type;
    normalized.mutability = mutability;
    return normalized;
  });
}

export default class Post extends Model {
  constructor(context) {
    super(context);

    this.collection = context.db.collection('post');
  }

  all({ limit = 10, offset = 0, search = null }) {
    const criteria = {};
    if (search) {
      criteria.$text = { $search: search };
    }

    return this.collection
      .find(criteria)
      .sort({ createdAt: -1 })
      .skip(offset)
      .limit(limit)
      .toArray();
  }

  async insert(doc) {
    if (!doc.title) {
      throw new Error('Post requires a title.');
    }
    const slug = await getUniqueSlug(this.collection, doc.title);
    const docToInsert = Object.assign({}, doc, {
      slug,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });
    docToInsert.contentState.entityMap = convertEntityData(docToInsert.contentState.entityMap);
    const id = (await this.collection.insertOne(docToInsert)).insertedId;
    return id;
  }

  async updateById(id, doc) {
    const docToUpdate = Object.assign({}, doc);
    docToUpdate.contentState.entityMap = convertEntityData(docToUpdate.contentState.entityMap);
    const ret = await this.collection.update(
      { _id: id },
      {
        $set: Object.assign({}, docToUpdate, {
          updatedAt: Date.now(),
        }),
      }
    );
    this.loader.clear(id);
    return ret;
  }
}
