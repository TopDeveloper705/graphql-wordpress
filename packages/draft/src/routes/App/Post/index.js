import React, { Component } from 'react';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import Helmet from 'components/Helmet';
import { settingsShape, socialSettingsShape } from 'types/PropTypes';
import NotFound from 'components/NotFound';
import Content from './Content';
import { Wrapper, Title } from './styled';

/* eslint-disable react/prop-types */

@graphql(
  gql`
    query PostQuery($slug: String) {
      post(slug: $slug) {
        id
        title
        slug
        contentState {
          ...Content_contentState
        }
        summary
        featuredMedia {
          destination
          ... on ImageUpload {
            crops {
              fileName
              width
            }
          }
        }
      }
    }
    ${Content.fragments.contentState}
  `,
  {
    options: ({ match: { params } }) => ({
      variables: { slug: params.slug },
    }),
  }
)
export default class PostRoute extends Component {
  static contextTypes = {
    settings: settingsShape,
    socialSettings: socialSettingsShape,
  };

  render() {
    const { data: { loading, error, post } } = this.props;

    if (error) {
      return <NotFound />;
    }

    if (loading && !post) {
      return null;
    }

    const { siteUrl } = this.context.settings;
    const { twitterUsername } = this.context.socialSettings;
    const postUrl = `${siteUrl}/post/${post.slug}`;

    let featuredImage = null;
    if (post.featuredMedia && post.featuredMedia.length > 0) {
      const media = post.featuredMedia[0];
      const crop = media.crops.find(c => c.width === 640);
      featuredImage = `${siteUrl}/uploads/${media.destination}/${crop.fileName}`;
    }

    return (
      <Wrapper>
        <Helmet>
          <title>{post.title}</title>
          <link rel="canonical" href={postUrl} />
          <meta property="og:type" content="article" />
          <meta property="og:title" content={post.title} />
          <meta property="og:description" content={post.summary} />
          <meta property="og:url" content={postUrl} />
          {featuredImage && <meta property="og:image" content={featuredImage} />}
          <meta name="twitter:card" value="summary_large_image" />
          {twitterUsername && <meta name="twitter:site" value={`@${twitterUsername}`} />}
          {twitterUsername && <meta name="twitter:creator" value={`@${twitterUsername}`} />}
          <meta name="twitter:title" content={post.title} />
          <meta name="twitter:description" content={post.summary} />
          <meta name="twitter:url" content={postUrl} />
          {featuredImage && <meta name="twitter:image" content={featuredImage} />}
        </Helmet>
        <Title>{post.title}</Title>
        <Content contentState={post.contentState} />
      </Wrapper>
    );
  }
}
