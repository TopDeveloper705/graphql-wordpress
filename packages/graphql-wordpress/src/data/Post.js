import { toGlobalId } from 'graphql-relay';
import fetchData from 'data/utils';

class Post {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return process.env.WP_POSTS_ENDPOINT || 'graphql/v1/posts';
  }

  static async collection(args = {}) {
    const { data: { body, headers } } = await fetchData(Post.getEndpoint(), {
      qs: args,
    });
    return {
      total: headers['x-wp-total'],
      items: body.map(item => Object.assign(new Post(), item)),
    };
  }
}

export default Post;
