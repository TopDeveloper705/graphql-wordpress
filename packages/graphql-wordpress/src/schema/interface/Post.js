import { GraphQLInterfaceType, GraphQLNonNull, GraphQLID, GraphQLList } from 'graphql';
import UserType from 'type/User';
import MediaType from 'type/Media';
import Meta from 'type/Meta';
import { slug, guid, link } from 'field/identifier';
import { date, modified } from 'field/date';
import { commentStatus, pingStatus } from 'field/status';
import { title, content, excerpt } from 'field/content';
import { type, template } from 'field/post';

const PostInterface = new GraphQLInterfaceType({
  name: 'PostInterface',
  fields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'Unique identifier for the object.',
    },
    ...date,
    ...guid,
    ...modified,
    ...slug,
    ...type,
    ...link,
    ...title,
    ...content,
    ...excerpt,
    ...commentStatus,
    ...pingStatus,
    ...template,
    meta: { type: new GraphQLList(Meta) },
    author: { type: UserType },
    featuredMedia: { type: MediaType },
  },
});

export default PostInterface;
