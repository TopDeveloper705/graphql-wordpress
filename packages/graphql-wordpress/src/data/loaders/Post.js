import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Post from 'data/Post';

// Dataloader expects IDs that can be read by the REST API

export default function getPostLoaders() {
  const endpoint = Post.getEndpoint();

  let slugLoader;
  const postLoader = new Dataloader(ids =>
    fetchData(endpoint, {
      qs: { include: ids, orderby: 'include', per_page: 100 },
    }).then(({ data: { body: posts } }) => {
      posts.forEach(post => {
        slugLoader.prime(post.slug, post);
      });
      return posts;
    })
  );
  slugLoader = new Dataloader(slugs =>
    fetchData(endpoint, {
      qs: { slug: slugs, orderby: 'slug', per_page: 100 },
    }).then(({ data: { body: posts } }) => {
      posts.forEach(post => {
        postLoader.prime(post.id, post);
      });
      return posts;
    })
  );

  return {
    load: async id => {
      const data = await postLoader.load(id);
      return data ? Object.assign(new Post(), data) : null;
    },
    loadBySlug: async slug => {
      const data = await slugLoader.load(slug);
      return data ? Object.assign(new Post(), data) : null;
    },
  };
}
