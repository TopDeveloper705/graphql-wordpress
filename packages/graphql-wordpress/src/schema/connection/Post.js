import { GraphQLString, GraphQLID, GraphQLInt, GraphQLBoolean } from 'graphql';
import {
  connectionArgs,
  connectionDefinitions,
  connectionFromArraySlice,
  fromGlobalId,
  cursorToOffset,
} from 'graphql-relay';
import PostType from 'type/Post';
import Post from 'data/Post';
import POST_ORDERBY from 'enum/PostOrderby';
import ORDER from 'enum/Order';

const { connectionType: PostConnection } = connectionDefinitions({
  nodeType: PostType,
});

export { PostConnection };

export default {
  type: PostConnection,
  args: {
    sticky: {
      type: GraphQLBoolean,
      description: 'Limit result set to items that are sticky.',
    },
    year: {
      type: GraphQLInt,
      description: 'Limit result set to items that are from a certain year.',
    },
    month: {
      type: GraphQLInt,
      description: 'Limit result set to items that are from a certain month.',
    },
    day: {
      type: GraphQLInt,
      description: 'Limit result set to items that are from a certain day.',
    },
    search: {
      type: GraphQLString,
      description: 'Limit results to those matching a string.',
    },
    author: {
      type: GraphQLID,
      description: 'Limit result set to items assigned to specific author.',
    },
    slug: {
      type: GraphQLString,
      description:
        'Limit result set to items with one or more specific slugs (value or comma-separated values).',
    },
    // value or comma-separated values
    category: {
      type: GraphQLString,
      description:
        'Limit result set to all items that have the specified term assigned in the categories taxonomy.',
    },
    categories: {
      type: GraphQLString,
      description:
        'Limit result set to all items that have the specified terms assigned in the categories taxonomy.',
    },
    tag: {
      type: GraphQLString,
      description:
        'Limit result set to all items that have the specified term assigned in the tags taxonomy.',
    },
    tags: {
      type: GraphQLString,
      description:
        'Limit result set to all items that have the specified terms assigned in the tags taxonomy.',
    },
    term: {
      type: GraphQLString,
      description:
        'Limit result set to all items that have the specified term assigned in the also-specified taxonomy.',
    },
    taxonomy: {
      type: GraphQLString,
      description:
        'Limit result set to all items that have the also-specified term assigned in this taxonomy.',
    },
    orderby: {
      type: POST_ORDERBY,
      description: 'Sort collection by object attribute.',
    },
    order: {
      type: ORDER,
      description: 'Order sort attribute ascending or descending.',
    },
    ...connectionArgs,
  },
  description: 'A list of results',
  resolve: (root, args) => {
    const connectionArguments = {};
    const params = Object.assign({}, args);
    if (typeof params.search !== 'undefined' && !params.search) {
      return null;
    }

    if (params.first) {
      params.per_page = params.first;
      connectionArguments.first = params.first;
    } else {
      params.per_page = 10;
    }

    if (params.after) {
      params.offset = cursorToOffset(params.after) + 1;
    }

    const setFilter = (arg, restArg = null) => {
      if (params[arg]) {
        if (!params.filter) {
          params.filter = {};
        }
        params.filter[restArg || arg] = params[arg];
        delete params[arg];
      }
    };

    if (params.taxonomy && params.term) {
      if (params.taxonomy === 'category') {
        setFilter('term', 'category_name');
      } else if (params.taxonomy === 'tag') {
        setFilter('term', 'tag');
      }
    } else {
      // query by category slug
      setFilter('category', 'category_name');
      // query by tag slug
      setFilter('tag');
    }

    // query by year
    setFilter('year');
    // query by month
    setFilter('month', 'monthnum');
    // query by day
    setFilter('day');

    ['categories', 'tags', 'author'].forEach(key => {
      if (params[key]) {
        params[key] = params[key].split(',').map(value => fromGlobalId(value).id);
      }
    });

    delete params.first;
    delete params.last;
    delete params.after;
    delete params.before;

    return Post.collection(params).then(({ items, total }) =>
      connectionFromArraySlice(items, connectionArguments, {
        arrayLength: total ? parseInt(total, 10) : items.length,
        sliceStart: params.offset || 0,
      })
    );
  },
};
