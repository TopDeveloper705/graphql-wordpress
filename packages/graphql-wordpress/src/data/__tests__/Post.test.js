import Post from 'data/Post';

const backupEnv = Object.assign({}, process.env);

afterAll(() => {
  process.env = backupEnv;
});

jest.mock('../utils', () =>
  jest.fn(() => ({
    data: {
      body: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }],
      headers: {
        'x-wp-total': 5,
      },
    },
  }))
);

describe('Test Post data access', () => {
  test('Get endpoint', () => {
    expect(Post.getEndpoint()).toMatchSnapshot();
  });

  test('Get backup endpoint', () => {
    delete process.env.WP_POSTS_ENDPOINT;
    expect(Post.getEndpoint()).toMatchSnapshot();
  });

  test('Load Post collection', async () => {
    const collection = await Post.collection();
    expect(collection).toMatchSnapshot();
  });
});
