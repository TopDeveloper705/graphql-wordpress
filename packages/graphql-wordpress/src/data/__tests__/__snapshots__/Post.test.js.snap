// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Post data access Get backup endpoint 1`] = `"graphql/v1/posts"`;

exports[`Test Post data access Get endpoint 1`] = `"graphql/v1/posts"`;

exports[`Test Post data access Load Post collection 1`] = `
Object {
  "items": Array [
    Post {
      "id": 1,
    },
    Post {
      "id": 2,
    },
    Post {
      "id": 3,
    },
    Post {
      "id": 4,
    },
    Post {
      "id": 5,
    },
  ],
  "total": 5,
}
`;
